Conveyor

barel13 opened the pull request:
# Conveyor

The main goal of the conveyor is to store the balls until we need to shoot them via the shooter.
<br> Another goal of the conveyor is to move (or convey) the balls to the shooter.

### Convey (Command)
The goal of this command is to move the balls from the funnel or to move the balls to the shooter.

### Balls
I created the balls object this way because knowing the amount of balls during the game is an integral part of the robot.

barel13 committed:
Initialize default classes for the conveyor

barel13 committed:
Create And implement DeadbandProximity class

barel13 committed:
Create Conveyor Motor

barel13 committed:
Create function for the conveyor

barel13 committed:
Add method to handle the balls

barel13 committed:
Create LoadConveyor command

barel13 committed:
Add safety feature to LoadConveyor command

barel13 committed:
Ignore .idea folder

barel13 committed:
Rename LoadConveyor command to Convey

barel13 committed:
Add documentation to the conveyor subsystem

barel13 committed:
Add support to a case when the conveyor take out ball threw the funnel

barel13 committed:
Remove the UnitModel from the conveyor

barel13 committed:
Fix the convey if statement

barel13 committed:
Add boolean to check the whether we shoot or store the balls

barel13 committed:
Fix build issue

katzuv commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,9 @@
+    public static final class Conveyor {
+        public static final int MOTOR = 0;
+        public static final boolean IS_MOTOR_INVERTED = false;
+        public static final int SHOOTER_PROXIMITY = 0;
```
```suggestion
        public static final int PROXIMITY_SENSOR = 0;
```
Why is the "shooter" needed?
Also, don't forget it's a sensor. And BTW, aren't you using a laser sensor?

barel13 replied:
We use Laser sensor, but I don't know how to call an object of this "type"

katzuv replied:
`LASER_SENSOR` is good.
However, I just realized that if you really use a laser sensor, you'll need to change some of the code because these are digital and not analog like proximity sensors.

barel13 replied:
@orel-david says that the sensor is analog

katzuv replied:
😢
That's unfortunate. I strongly recommend using laser beam sensors, they are simpler to work with and more accurate IMO.
Anyway, if you are using a proximity sensor, why https://github.com/Galaxia5987/robot-2021/commit/5cb1adaa02fe495155a7d02475ca36c6fb01ee63?

barel13 replied:
We are using the Laser Beam, We are going to connect it (somehow, don't ask me how) to an analog sensor

katzuv replied:
There is some confusion here. Usually when talking about "laser beam sensors" you talk about a pair of items, where one launches a laser beam and the other receives it ([example](https://www.aliexpress.com/item/4000060455626.html)). The output of the sensor is **digital**: "0" if the sensor doesn't receive the beam, "1" otherwise. So when an object passes through the beam and blocks it, you can know that.

A "proximity sensor" is a sensor that consists of one item that launches a laser beam and receives its reflection. Based on the [time of flight](https://en.wikipedia.org/wiki/Time_of_flight) (ToF), it outputs a varying voltage which is proportional to the time it took the beam to return to the sensor. This sensor is **analog** because it outputs a range of values and not just a pair of them like a digital sensor.

So which one are you using in the conveyor?
> somehow, don't ask me how

I think it's good knowledge to have – you program the subsystem, you should know how it works.

barel13 replied:
I know how laser beam and proximity sensor works, I told you that they are planning to buy something that is equivalent in the correctness of the laser beam (even-though most likely it won't happen) ,or they somehow going to connect it to an "translator"

katzuv replied:
You see, I like writing texts.
> I know how laser beam and proximity sensor works

So although I realized that while I was writing the explanation, I wanted to finish it anyway 😝.
> I told you [...] something that is equivalent in the correctness of the laser beam

I'm not sure what that means, but your [comment](https://github.com/Galaxia5987/robot-2021/pull/9#discussion_r566653975) doesn't mention anything like this.
> I'm sorry, but I still don't understand what type the sensor is – analog or digital?

barel13 replied:
I told you the second thing, and for your 2nd question, just ask @orel-david, he can explain it better

katzuv replied:
Sure, I'll wait for his response. But I think it's important you'll know how your subsystem works.

orel-david replied:
We havn't bought the sensor yet and some of the sensors we looked at were analog, so for now it is fine if the code refers to them as analog ,because even if they are digital they would return 5v or 0 so the deadband approach would still work

katzuv replied:
I see.
> because even if they are digital they would return 5v or 0 so the deadband approach would still work

True, but treating the sensor as digital will be simpler.

katzuv commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -11,7 +11,6 @@
 import edu.wpi.first.wpilibj.XboxController;
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.button.JoystickButton;
-import frc.robot.subsystems.ExampleSubsystem.ExampleSubsystem;
```
This is good, but not a part of this PR. Please open another PR where you'll delete the example files.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,97 @@
+import static frc.robot.Constants.Conveyor.*;
+
+public class Conveyor extends SubsystemBase {
+    private static int balls = Constants.Conveyor.INITIAL_BALLS_AMOUNT;
+
+    private final TalonFX motor = new TalonFX(Ports.Conveyor.MOTOR);
```
Why is `balls` static? You have only one instance of this class anyway.

barel13 replied:
I use it because I don't think that the `balls` are specific to the conveyor, the balls is being used in most of the subsystem of the robot.
If you want, I can move it to the `Robot` class, because the balls do not belong to a specific subsystem...

katzuv replied:
I think it's OK to leave `balls` here, but I'm still not convinced why it should be static.
Where do you need to access the number of Power Cells outside of this subsystem context? In the intake, to prevent intaking if the number of Power Cells is the allowed maximum?

barel13 replied:
Exactly

katzuv replied:
OK, so instead, export a method from the `Conveyor` class which says whether the amount of Power Cells is at the allowed maximum.
And in any case, `balls` doesn't need to be `static`, and all the `synchronized` keywords are redundant.

barel13 replied:
first of all, you didn't provide any reason for not using `synchronized` or `static`.
Second of all, No matter what, I still need to update the amount of balls (for example, add ball via the funnel and remove ball via the shooter), so I need to keep the `synchronized` keyword.

katzuv replied:
> first of all, you didn't provide any reason for not using `synchronized` or `static`.

That's on you. You add code, you need to justify its existence.
> No matter what, I still need to update the amount of balls

True, but how does that imply
> I need to keep the `synchronized` keyword

?
You need `synchronized` when you write multi-threaded code (which you don't), and `static` when you have multiple instances of a class (which you don't, as subsystems are [singletons](https://en.wikipedia.org/wiki/Singleton_pattern)).

barel13 replied:
> That's on you. You add code, you need to justify its existence.

And you need to show why we don't need to update the value.
I will delete (some of) the `synchronized`, but not because what you said, because you didn't give me any reason that I'm wrong.

> `static` when you have multiple instances of a class

not only, when the variable is not (directly) related to the class. The `balls` variable isn't related only to the conveyor.
I told you, if you want, I can move it to the Robot class

katzuv replied:
> but not because what you said

At first, I thought whether I should feel offended.
> I will delete (some of) the `synchronized`, but not because what you said, because you didn't give me any reason that I'm wrong.

But I did (https://github.com/Galaxia5987/robot-2021/pull/9#discussion_r566845009):
> You need `synchronized` when you write multi-threaded code (which you don't)
<br>

> And you need to show why we don't need to update the value.

Again, I've never said you don't need to update the amount of Power Cells the conveyor holds:
> > No matter what, I still need to update the amount of balls
>
> True
<br>

> not only, when the variable is not (directly) related to the class

Could you please provide an example? I've never considered that to be a good design idea. A class should only contain members that are related to it.
> The `balls` variable isn't related only to the conveyor. I told you, if you want, I can move it to the Robot class

You're right, I rejected that idea at first. After all, the conveyor stores the Power Cells, and the only subsystem other than it that cares about the number of them is, maybe, the intake. But that's more than one subsystem, which justifies moving it to the `Robot` class…
_However_, while writing the above paragraph, I thought again it's not a good idea – by storing the counter in the `Robot` class, you let every other component in the code change it. But it should only be modified by the conveyor because it holds the sensors responsible to count the Power Cells.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,97 @@
+        motor.configStatorCurrentLimit(STATOR_CURRENT_LIMIT, Constants.TALON_TIMEOUT);
+
+        motor.enableVoltageCompensation(true);
+        motor.configVoltageCompSaturation(12);
```
Magic number.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+        if (toShooter) {
+            if (Conveyor.getBallsAmount() > 0)
+                conveyor.setPower(power);
+            // else TURN ON LEDS
```
Add a `TODO:` so you won't forget to implement that.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public void execute() {
+        if (toShooter) {
+            if (Conveyor.getBallsAmount() > 0)
+                conveyor.setPower(power);
```
```suggestion
            if (Conveyor.getBallsAmount() > 0) {
                conveyor.setPower(power);
            }
```
Always wrap blocks with brackets.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public void execute() {
+        if (toShooter) {
+            if (Conveyor.getBallsAmount() > 0)
```
I'm not sure this condition is desirable. You want to be able to shoot even if the sensor malfunctions.

barel13 replied:
Can you please elaborate?

katzuv replied:
I see you're using my words againat me 🙃
Say there's a problem in the sensor and the Power Cells counter says there are `-5` Power Cells in the robot, while in fact there are `2`. This condition won't allow shooting.

you could argue that if the sensor malfunctions, other features won't work (I believe, for example, you'll prevent intaking Power Cells if there are 5 in the robot). However, in that specific case you won't lose anything from running the conveyor, even if it's empty.

barel13 replied:
You'll lost some power (from the battery), and if you really want, we can always add an "override" mode, that override almost, if not every check. Anyway, I don't think I really should add the checks here, (in my opinion) we need to create the "safety features" in another branch. I just did it because I knew that you would (probably) give me a comments about that I don't check some of the cases

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+    public Convey(Conveyor conveyor, double power, boolean toShooter) {
+        this.conveyor = conveyor;
+        this.power = power;
+        this.toShooter = toShooter;
```
Why not just a negative `power` if you want to bring Power Cells back inside?

barel13 replied:
I don't want to bring a power cell back to the conveyor, I might want to get the power cell out of the robot. I don't want to use this mechanism because I might have a different `power` to apply when we use it in conjunction with the funnel

katzuv replied:
> I don't want to bring a power cell back to the conveyor, I might want to get the power cell out of the robot.

Aren't those two the same? You eject Power Cells by driving the conveyor down.

>  I might have a different `power` to apply

I'm not sure I understand where the problem is. If you need another power, supply another to the command.

katzuv replied:
You didn't answer my questions...

barel13 replied:
You didn't answer mine...
Either way, I'm going to split the command into different commands, because this command handles a lot of situation that aren't related to each other

katzuv replied:
> You didn't answer mine...

Did I miss anything? All question marks in this conversation are mine 😁.

barel13 replied:
In this conversation yes, but not in this PR
I asked [this](https://github.com/Galaxia5987/robot-2021/pull/9#pullrequestreview-566206941) 17 days ago...

katzuv replied:
Sorry, missed that. [As you asked](https://github.com/Galaxia5987/robot-2021/pull/8#commitcomment-45818341), it's a bit hard to follow comments written on the PR. Please tag if you do, so I'll receive a notification.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+                conveyor.setPower(power);
+            // else TURN ON LEDS
+        } else
+            if (Conveyor.getBallsAmount() < Constants.Conveyor.MAX_BALLS_AMOUNT)
```
Could you please explain why is this check necessary?

barel13 replied:
It's for the case we want to move the balls to the conveyor, I know I need to rename the `boolean`, I will change it later

katzuv commented on src/main/java/frc/robot/utils/DeadbandProximity.java:
```diff
@@ -0,0 +1,82 @@
+ * the sensor switches.
+ */
+public class DeadbandProximity extends AnalogInput {
+    private final double lostVoltage;
+    private final double senseVoltage;
```
"lost" is unclear. I think using "seen" and "unseen" are better and it's also clearer the two values are related.

barel13 replied:
I disagree. In a technical view, the sensor can't "see", he doesn't have eyes, he only can sense.

katzuv replied:
In a technical view, I'm not sure the sensor "senses", it just sends a varying voltage.
What's important is how we humans refer to it, not the technical terms. Anyway, "lost" still doesn't fit here IMO.

barel13 replied:
First of all, if a sensor can't sense, it's really misleading.
Secondly, why lost doesn't fit? it's pretty clear that it refers to the minimal voltage that the sensor sense, i.e. when the object is lost

katzuv replied:
It fits, but "seen"/"unseen" or "sensed"/"unsensed" fit better, because it's clear the two flags are related to each other.
> First of all, if a sensor can't sense, it's really misleading.

If a sensor can't see because it doesn't have eyes, how can it sense if it doesn't have senses? The technical terms don't matter, only the terms we as humans use matter.

barel13 replied:
I can use `ignoranceVoltage`, I don't agree that those words, they don't "tell" what is the real purpose of the variables

katzuv replied:
They _do_ tell the real purpose: the outputted voltage when the sensor sees a Power Cell, and the outputted voltage when it doesn't see one.

barel13 replied:
the sensor **can't** see, I'm not going to lie in my code, like in the shooter you fought with me to use `estimateVelocityFromDistance` just because it's more "correct", I'm not going to allow to change the name of the `lostVoltage` to another name as long as the name isn't more correct from my variable's name

katzuv replied:
> you fought with me

I don't fight. I discuss. 🌷
> I'm not going to lie in my code

Not using the technical term ≠ lying.

Just gonna put here a few articles which also say that sensors see (emphasis mine):
- "Radar sensor for the home which can '_see_' through walls" ([source](https://www.bbc.com/news/av/technology-38608135))
- "it’s important to make sure each sensor has a clear _line of sight_" (sight ≡ see, [source](https://www.safewise.com/blog/effective-placement-can-prevent-a-burglary-where-your-sensors-should-go/))
- "What a Sensor _Sees_" ([source](https://www.te.com/global-en/trends/electric-vehicles-connected-transportation/what-a-sensor-sees-in-automotive.html))

This is my last comment on this matter unless you want to further discuss it.

katzuv commented on src/main/java/frc/robot/utils/DeadbandProximity.java:
```diff
@@ -0,0 +1,82 @@
+import edu.wpi.first.wpilibj.AnalogInput;
+
+/**
+ * This class is a modification of the regular proximity class, which adds a deadband, to insure that the sensor
```
```suggestion
 * This class is a modification of the regular proximity class, which adds a deadband, to ensure that the sensor
```

barel13 replied:
> The cynic would be mostly right, at least for much of the history of English, as for hundreds of years insure and ensure were simply spelling variants, and had no more difference between them than theatre and theater. ([Miriam-Webster dictionary](https://www.merriam-webster.com/words-at-play/how-to-use-insure-vs-ensure-vs-assure))

katzuv replied:
The key part is "for much of the history of English". As written later in the article you sent:
> Many usage guides have suggested restricting the use of _insure_ to financial matters, and employing _ensure_ in general uses where you mean “to make sure, certain, or safe.”

[Grammarly](https://www.grammarly.com/blog/assure-ensure-insure/) says the same:
> To **ensure** something is to make sure it happens—to guarantee it.
To **insure** something or someone is to cover it with an insurance policy.

katzuv commented on src/main/java/frc/robot/utils/DeadbandProximity.java:
```diff
@@ -0,0 +1,82 @@
+     * We use this method to ensure that the proximity doesn't toggle rapidly because of sensor noise.
+     */
+    public void updateState() {
+        boolean lastState = objectSensed;
```
```suggestion
        final boolean lastState = objectSensed;
```

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,97 @@
+        shooterProximity.updateState();
+        funnelProximity.updateState();
+
+        boolean movingUp = getPower() >= 0;
```
You're using `getPower` only here. I think it's better to replace `getPower` with a method that returns exactly what you're looking for.

katzuv reviewed (changes requested):
Could you please stop pushing all the should-be-ignored files? It **really** clutters the diffs.

barel13 reviewed (commented):
I will apply the changes later

barel13 committed:
Change the purpose of the getPower method

barel13 committed:
make lastState boolean final

barel13 committed:
Add braced to convey command, and add TODO

barel13 committed:
Use constants instead of "magic numbers"

adamv17 commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,96 @@
+     * @return whether the power applied by the motor is upward.
+     */
+    public boolean isMovingUp() {
+        return motor.getMotorOutputPercent() >= 0;
```
I think this has to be greater than 0 because the ball could be moving down and then stop at the edge of sight this would add the ball from the count, instead of just minus. This would only happen if the balls could slip slightly in the system, which is probable. This situation does not occur when the balls are moving up because they would be counted again. This actually counteracts if the balls are moving up and then slip down when the conveyor stops because then the sensor would minus and not add extra.

barel13 replied:
Do you talk about a case when a ball enters the conveyor from the funnel, or when we try to shoot?

adamv17 replied:
I think it would be true for both.

barel13 replied:
I agree but not (just) because of these reasons, I agree because if another robot hits us or the robot stops at once, then the balls might move.

adamv17 replied:
Yeah that is plausible as well

adamv17 commented on src/main/java/frc/robot/utils/DeadbandProximity.java:
```diff
@@ -0,0 +1,82 @@
+     * Update the state of the proximity.
+     * An object considered as sensed when the value goes above the {@link #senseVoltage}.
+     * An object considered as lost when the value goes below the {@link #lostVoltage}.
+     * We use this method to ensure that the proximity doesn't toggle rapidly because of sensor noise.
```
Consider using passive form - "This method ensures"

adamv17 commented on src/main/java/frc/robot/utils/DeadbandProximity.java:
```diff
@@ -0,0 +1,82 @@
+     * Get whether the proximity sense an object.
+     * If you wish to check whether the proximity lost the object, use {@link #isObjectAway()} instead.
+     *
+     * @return whether the proximity sense a object.
```
*sensed an

barel13 committed:
Fix documentation

barel13 reviewed (commented):
Question: does the method `getMotorOutputPercent()` returns the value based on the velocity and the maximal velocity, or just by the power the the motor applies? if it's the second option, I need to change the `isMovingUp()` method

barel13 committed:
Change isMovingUp logic

orel-david commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,96 @@
+        shooterProximity.updateState();
+        funnelProximity.updateState();
+
+        if (!shooterProximity.hasObjectSensed() && shooterProximity.hasStateChanged() && isMovingUp() ||
+                !funnelProximity.hasObjectSensed() && funnelProximity.hasStateChanged() && !isMovingUp()) {
+            Conveyor.removeBall();
+        }
+
+        if (funnelProximity.hasObjectSensed() && funnelProximity.hasStateChanged() && isMovingUp()) {
+            Conveyor.addBall();
+        }
```
Add some documentation

orel-david commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public void execute() {
+        if (toShooter) {
+            if (Conveyor.getBallsAmount() > 0) {
+                conveyor.setPower(power);
+            }// TODO: else TURN ON LEDS
+        } else if (Conveyor.getBallsAmount() < Constants.Conveyor.MAX_BALLS_AMOUNT) {
+            conveyor.setPower(power);
```
Don't you want to add a condition that the shooter is ready and the swerve is aimed at the target?

barel13 replied:
I don't think we should do it here, it's more about integration

katzuv replied:
I agree with Barel, IMO every command should require up to one subsystem.

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/commands/Convey.java:
```diff
@@ -0,0 +1,43 @@
+        if (toShooter) {
+            if (Conveyor.getBallsAmount() > 0) {
+                conveyor.setPower(power);
+            }// TODO: else TURN ON LEDS
```
LEDs for what? Signaling what exactly?

barel13 replied:
That we (might) don't have balls inside the conveyor.

barel13 committed:
Rename DeadbandProximity to DeadbandLaserSensor

barel13 committed:
Add another port to the funnel sensor

barel13 committed:
Add documentation to the sensor state-machine

barel13 committed:
Refactor Convey to FeedShooter

barel13 committed:
Create LoadConveyor command

barel13 committed:
Add methods to handle more appropriately the use of the balls

barel13 committed:
Implement LoadConveyor command

barel13 committed:
Add documentation to LoadConveyor command

katzuv commented:
> Question: does the method `getMotorOutputPercent()` returns the value based on the velocity and the maximal velocity, or just by the power the the motor applies? if it's the second option, I need to change the `isMovingUp()` method

What's the difference between the two options?
@barel13

barel13 commented:
There is no difference now, When I wrote it there was a difference

katzuv commented:
> There is no difference now, When I wrote it there was a difference

Was anything changed in the Talon SRX's firmware in the last 17 days?
Anyway, if there is no difference, why did you ask me to answer your not-answered question? 
(Not angry, just curious)

barel13 committed:
Use AtomicInteger instead of int

Just to avoid the case that we call some related function in parallel when we change the amount of balls, for example, we check whether the conveyor is full and we have 4 balls, and by the same time we insert another ball, what would be the value from the `isConveyorFull()`

orel-david commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,108 @@
+     * @return whether the power applied by the motor is upward.
+     */
+    private boolean isMovingUp() {
+        return motor.getMotorOutputPercent() > 0;
```
Because it is talonFX and you have a built in encoder it would be more accurate if you check the speed and not the precent output

katzuv replied:
You don't really need accuracy, you just want to know the direction the conveyor is moving in.
However, maybe you should have another method that returns whether the conveyor is moving down because this method returns `false` if the conveyor doesn't move at all.

barel13 replied:
I don't agree with you @katzuv, if we are not moving up, the balls still can slip and it won't handle it

katzuv replied:
> it won't handle it

I'm not sure what you mean, please be more specific.
> if we are not moving up, the balls still can slip

If Power Cells slip when the conveyor is static, they will slip when it moves, too. The friction between Power Cells and the conveyor is the same, regardless of whether it's moving or not.

barel13 replied:
> The friction between Power Cells and the conveyor is the same, regardless of whether it's moving or not.

?

> I'm not sure what you mean, please be more specific.

I'm talking about the case when the motor doesn't apply power and the balls slip

katzuv replied:
The force that drives the Power Cells up in the conveyor is static friction. IIRC from my physics classes, the formula for static friction between two objects is <img src="https://render.githubusercontent.com/render/math?math=F_s = \mu_s N">. This force's magnitude is not dependant on the velocity of any of the two objects.
Thus, if Power Cells slip when the conveyor is static, they'll slip when it moves, too.

BTW, I'm not sure how we came to discuss this, because my original comment was:
> maybe you should have another method that returns whether the conveyor is moving down because this method returns `false` if the conveyor doesn't move at all.
https://github.com/Galaxia5987/robot-2021/pull/9#discussion_r567111457

orel-david commented on src/main/java/frc/robot/subsystems/conveyor/commands/FeedShooter.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public boolean isFinished() {
+        return false;
```
Maybe add that if the conveyor is empty than the function should end

barel13 replied:
?

orel-david commented on src/main/java/frc/robot/subsystems/conveyor/commands/LoadConveyor.java:
```diff
@@ -0,0 +1,38 @@
+    private final Conveyor conveyor;
+    private final double power;
+
+    public LoadConveyor(Conveyor conveyor, double power) {
```
Notice that you want to move the conveyor the minimal distance to enter each power cell in order to optimize our space usage

orel-david commented on src/main/java/frc/robot/subsystems/conveyor/commands/LoadConveyor.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void execute() {
+        if (!Conveyor.isConveyorFull()) {
```
Maybe use the funnel sensor to check if the ball entered

barel13 replied:
How would it help in a case we are planning to load more than one ball?

orel-david replied:
You would move the conveyor while there is a ball at the entrence

katzuv commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,108 @@
+import static frc.robot.Constants.Conveyor.*;
+
+public class Conveyor extends SubsystemBase {
+    private static final AtomicInteger balls = new AtomicInteger(Constants.Conveyor.INITIAL_BALLS_AMOUNT);
```
Cool, I didn't know about `AtomicInteger`. But shouldn't the `Atmoic*` stuff be used when you have multiple threads that can change the variable?

katzuv replied:
You over-complicate the code and I see no reason for that.
The code is single-threaded, so you are at no risk `balls`'s value will be wrong. Thus, using `synchronized` or `AtomicInteger` is unnecessary.
Moreover, you don't need `static` because this class is a singleton.

In the code for _Asteria_, `balls` wasn't static nor `synchronized` or `AtomicInteger` were used, and the code worked perfectly. In fact, [it was you who've written it](https://github.com/Galaxia5987/robot-2020/blob/0c6325625ff0e0f5544c819be2e98ad6173b071e/src/main/java/frc/robot/subsystems/conveyor/Conveyor.java#L24)…

This is my last comment on the matter unless you wish to discuss it more.

barel13 replied:
You don't see a reason, but I see.
Example, What if we run the `FeedShooter` command, and in the beginning we have 4 balls, and another ball enter into the conveyor (so we need to increment the amount of balls) and by the **same** time, a ball is being taken out, now the amount of balls can be either 3,4 or 5, and the data is lost, and btw, it didn't work as we expected, so we didn't use it in the competition....

katzuv replied:
It will still be OK.
First of all, the chance this scenario will happen is very rare – it means you're intaking and shooting at the same time.
But even if you do so it will be fine. Let's take your example where `balls == 4`, and a Power Cell enters the conveyor while another exists through the shooter. Let's look at the code inside `periodic()`:
https://github.com/Galaxia5987/robot-2021/blob/3332132f23f8062f4c06044016b6644f5b959458/src/main/java/frc/robot/subsystems/conveyor/Conveyor.java#L98-L106
The condition at line 98 will evaluate to `true`, so line 100 will decrement `balls` to `3`.
The condition at line 104 will evaluate to `True`, so line 106 will increment `balls` to `4`.

Again, your code is synchronous. Even if something happens "by the **same** time", it will happen one after the other.

BTW, after looking at the code referenced above, I realized that if a Power Cell _exists_ through the intake while another is being shot, `balls` really will hold a wrong value (which is another rare scenario). See why?
That would happen even if you used `synchronized`, but fortunately, there's an easy fix.

barel13 replied:
Congratulation, You passed the test.
And now the real problem IMO:
Do you think that the `periodic()` function stops when we start running a command? Answer: No, the `periodic()` function runs and doesn't care about the commands. How is that, that the function still running while the commands run? They run the commands and the `periodic()` in other threads...
BTW, I don't need to make everything thread safe, only the `isConveyor...` functions, because only they run on other threads.

P.S, even if the scenario is _rare_ it still can happen, and I should think about it and find a way to handle it.

katzuv replied:
> Congratulation, You passed the test.

![image](https://media.tenor.com/images/d936610c7ab637c0c8c7bed796e6a202/tenor.gif)
> Do you think that the periodic() function stops when we start running a command?

I don't know.
> Answer: No

Have you looked into WPILib internals?
> How is that, that the function still running while the commands run?

Maybe it doesn't.

> How is that, that the function still running while the commands run? They run the commands and the periodic() in other threads...

Well, you might be right. I don't know how many threads are running in WPILib. But it doesn't matter, because `Shooter.periodic()` is the only function that changes `balls` and there is no reason for it to be run in more than one thread.
Also, it might still work single-threaded. Your computer seems to be multi-tasking, although you have i.e. eight cores in your CPU, but hundreds of processes that run in parallel. The roboRIO's CPU is dual-core ([source](https://www.ni.com/pdf/manuals/375275a.pdf), pg. 1), so even if every command and every subsystem's `periodic()` runs in another thread, only two of them can run at the same time. So it's very probable `periodic()` (sometimes, at least) gives way to another command to run.

> even if the scenario is rare it still can happen, and I should think about it and find a way to handle it.

I agree.

barel13 replied:
> only two of them can run at the same time

you know that cores != amount of threads, right?

and I just saw the implementation of periodic and it **still** runs when commands run...

I will remove the synchronized just because I saw the implementation.


>But it doesn't matter, because Shooter.periodic() is the only function that changes balls and there is no reason for it to be run in more than one thread.

How the hack is it related? if the balls are being called by more than 1 thread then you need to use `synchronized` (as I thought that periodic runs on another thread, different from the commands...)

katzuv replied:
> you know that cores != amount of threads, right?

I'm not talking about the total number of threads, but the total number of **simultaneously running** threads. Unless your CPU uses [Intel Hyper-Threading](https://en.wikipedia.org/wiki/Hyper-threading) or similar technology (in which case you'll count "logical cores"), I don't see a scenario where you can run more threads than the number of cores.
> and I just saw the implementation of periodic and it **still** runs when commands run...

They both might be scheduled to run, but when they actually run is up to the operating system to decide.

> How the hack is it related? if the balls are being called by more than 1 thread then you need to use `synchronized`

No, only if `balls` is **changed** by more than one thread. A race condition, which is the situation you describe here, can happen only if multiple threads try to change shared data at the same time.

barel13 committed:
Single-thread

katzuv reviewed (approved):
As with #8, I'll merge this pull request so you can get going in building your robot.
That was a great PR to go over, thank you because I learned a few stuff in Java (all the `synchronized` stuff) 🙂.

adamv17 commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,106 @@
+        shooterSensor.updateState();
+        funnelSensor.updateState();
+
+        // An object is being taken out either from the shooter or from the funnel
```
.

barel13 replied:
gg

adamv17 commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,106 @@
+            Conveyor.removeBall();
+        }
+
+        // An object is being inserted into the conveyor
```
.

barel13 replied:
gg

adamv17 commented on src/main/java/frc/robot/subsystems/conveyor/Conveyor.java:
```diff
@@ -0,0 +1,106 @@
+        funnelSensor.updateState();
+
+        // An object is being taken out either from the shooter or from the funnel
+        if (!shooterSensor.hasObjectSensed() && shooterSensor.hasStateChanged() && isMovingUp() ||
```
What happens when the ball passes the shooter senser and is counted but is not actually shot? Will the conveyor add it back?

adamv17 replied:
sensor

barel13 replied:
Can't happen!

barel13 committed:
Fix documentation

barel13 committed:
Alter FeedShooter finish condition

barel13 committed:
General fixes

barel13 committed:
Create MinimizeConveyor command

barel13 committed:
Use 2 proximity sensors in the DeadbandProximity

orel-david committed:
Fix

orel-david committed:
Add color sensor

orel-david committed:
Invert motor
